Add Vagrant Configuration

rajadain opened the pull request:
## Overview

This project is really slow to work with on Mac OS due to issues with running Docker on Mac (see #43).

This PR adds a Vagrant environment that can be used on non-Linux hosts to provide a Linux-like environment in which Docker runs natively and much faster.

On `develop`, running the `update` script can take ~20 minutes. Inside Vagrant, it takes ~4.5 minutes.

### Checklist

- [x] Update README with appropriate instructions for non-Linux environments
- [x] Description of PR is in an appropriate section of `CHANGELOG.md` and grouped with similar changes, if possible

## Testing Instructions

* Checkout this branch and `vagrant up`
* SSH in to the Vagrant VM `vagrant ssh`
* Run `./scripts/update`
  - [x] Ensure it runs faster than before 
* Run `./scripts/server`
  - [x] Ensure the server runs correctly http://localhost:3003/
* Make a small change in some front-end file
  - [x] Ensure it is picked up and hot-reload / TypeScript checking works as expected
* Make an account, then load dev data `./scripts/load-dev-data`
  - [x] Ensure the import succeeds
* Do other development-y things
  - [x] Ensure the app works in general and all existing development workflows work

kevinearldenny commented:
Hmmm, I'm getting some output that looks a little concerning after running `./scripts/load-dev-data` then `./scripts/server`:

```
server_1    | (node:28) UnhandledPromiseRejectionWarning: Error: ENOSPC: System limit for number of file watchers reached, watch '/home/node/app/server/src/templates'
server_1    |     at FSWatcher.start (internal/fs/watchers.js:210:26)
server_1    |     at Object.watch (fs.js:1444:11)
server_1    |     at createFsWatchInstance (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:118:15)
server_1    |     at setFsWatchListener (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:165:15)
server_1    |     at NodeFsHandler._watchWithNodeFs (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:330:14)
server_1    |     at NodeFsHandler._handleDir (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:551:19)
server_1    |     at processTicksAndRejections (internal/process/task_queues.js:97:5)
server_1    |     at async NodeFsHandler._addToNodeFs (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:600:16)
server_1    |     at async /home/node/app/server/node_modules/chokidar/index.js:435:21
server_1    |     at async Promise.all (index 0)
server_1    | (node:28) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
server_1    | (node:28) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
server_1    | (node:28) UnhandledPromiseRejectionWarning: Error: ENOSPC: System limit for number of file watchers reached, watch '/home/node/app/server/src/templates/passwordreset.hbs'
server_1    |     at FSWatcher.start (internal/fs/watchers.js:210:26)
server_1    |     at Object.watch (fs.js:1444:11)
server_1    |     at createFsWatchInstance (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:118:15)
server_1    |     at setFsWatchListener (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:165:15)
server_1    |     at NodeFsHandler._watchWithNodeFs (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:330:14)
server_1    |     at NodeFsHandler._handleFile (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:360:23)
server_1    |     at NodeFsHandler._addToNodeFs (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:620:21)
server_1    | (node:28) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
server_1    | (node:28) UnhandledPromiseRejectionWarning: Error: ENOSPC: System limit for number of file watchers reached, watch '/home/node/app/server/src/templates/verify.hbs'
server_1    |     at FSWatcher.start (internal/fs/watchers.js:210:26)
server_1    |     at Object.watch (fs.js:1444:11)
server_1    |     at createFsWatchInstance (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:118:15)
server_1    |     at setFsWatchListener (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:165:15)
server_1    |     at NodeFsHandler._watchWithNodeFs (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:330:14)
server_1    |     at NodeFsHandler._handleFile (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:360:23)
server_1    |     at NodeFsHandler._addToNodeFs (/home/node/app/server/node_modules/chokidar/lib/nodefs-handler.js:620:21)
server_1    | (node:28) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 3)
```

Everything still seems to work okay, but thought it was worth noting these since I can't remember seeing them when using Docker without Vagrant.

That being said, `./scripts/update` does seem to be significantly faster here. While builds with Docker would typically take 15-20 minutes, builds with Vagrant appear to consistently clock under 5 minutes 😄

kevinearldenny commented:
Another potentially concerning output from attempted hot reload after making a small change to a file in `client`:

```
server_1    | query: SELECT "RegionConfig"."id" AS "RegionConfig_id", "RegionConfig"."name" AS "RegionConfig_name", "RegionConfig"."country_code" AS "RegionConfig_country_code", "RegionConfig"."region_code" AS "RegionConfig_region_code", "RegionConfig"."s3_uri" AS "RegionConfig_s3_uri", "RegionConfig"."version" AS "RegionConfig_version", "RegionConfig"."hidden" AS "RegionConfig_hidden" FROM "region_config" "RegionConfig"
server_1    | [Nest] 40   - 05/04/2021, 3:58:23 PM   [TopologyService] downloaded data for s3URI s3://global-districtbuilder-dev-us-east-1/regions/US/WI/2021-02-08T15:19:22.145Z/
client_1    | Compiling...
server_1    | [Nest] 40   - 05/04/2021, 3:59:11 PM   [TopologyService] downloaded data for s3URI s3://global-districtbuilder-dev-us-east-1/regions/US/PA/2021-02-08T18:07:18.957Z/
server_1    | Killed
```

rajadain commented:
The `Error: ENOSPC: System limit for number of file watchers reached` issue has been fixed in 6fd467f.

rajadain commented:
The Nest `Killed` issue should be fixed in abf339c.

To test these two changes, you'll first have to reprovision your VM:

```
$ vagrant reload --provision
```

Could you take another look after doing that please?

rbreslow commented on deployment/ansible/roles.yml:
```diff
@@ -0,0 +1,8 @@
+  version: 2.0.0
+
+- src: azavea.docker
+  version: 5.0.0
```
This version bump may require changes to the `group_vars`, but I think we should do it.

```suggestion
  version: 6.0.0
```

See: https://github.com/azavea/ansible-docker/blob/develop/CHANGELOG.md#600

rajadain replied:
We currently have these in `group_vars`:

```
docker_version: "5:19.*"
docker_compose_version: "1.26.*"
```

What should they be upgraded to?

rbreslow replied:
I took a closer look at the changes, and I don't think you need to update anything. I'd verify things work after the bump.

rajadain replied:
They do indeed! Thanks for checking.

rbreslow commented on deployment/ansible/roles/district-builder.docker/tasks/main.yml:
```diff
@@ -0,0 +1,16 @@
+        append=yes
+  with_items: "{{ docker_users }}"
+
+- name: Increase file watching limit for Node
+  sysctl: name=fs.inotify.max_user_watches
+          value=524288
+          sysctl_set=yes
+          state=present
+          reload=yes
```
How did you decide on this value?

rajadain replied:
Based on https://github.com/PublicMapping/districtbuilder/pull/729#issuecomment-832394371 and https://gist.github.com/ntamvl/7c41acee650d376863fd940b99da836f. I cannot trace the true origin of that number. A lot of folks refer to http://blog.sorah.jp/2012/01/24/inotify-limitation which is down. But if you google for "inotify limitation 524288" a lot of results come up.

As to why we have to do this, it's because `tsc` wants to watch all files, even those within `node_modules`.

rbreslow replied:
It's worth perusing https://github.com/torvalds/linux/commit/92890123749bafc317bbfacbe0a62ce08d78efb7. Here, you can see that `inotify.max_user_watches` is proportional to the amount of addressable memory available. I wrote a small script in an attempt to demonstrate the formula:

<details>
<summary>Details</summary>

```c
#include <stdio.h>
#include <sys/sysinfo.h>
#include <sys/param.h>

// cat /usr/lib/modules/5.12.9-arch1-1/build/arch/x86/include/asm/page_types.h | grep PAGE_SHIFT
#define PAGE_SHIFT 12

// With my x86-64 config, the size of xfs_inode, proc_inode and
// inotify_inode_mark is 1680 bytes. The estimated INOTIFY_WATCH_COST is
// 1760 bytes.
#define INOTIFY_WATCH_COST 1760

int main()
{
    unsigned long watches_max;
    struct sysinfo si;

    sysinfo(&si);

    watches_max = (((si.totalram - si.totalhigh) / 100) << PAGE_SHIFT) /
            INOTIFY_WATCH_COST;
    // Substitute MIN and MAX macros for clamp
    watches_max = MIN(1048576U, MAX(watches_max, 8192U));


    printf("watches_max: %d\n", watches_max);

    return 0;
}
```

</details>

I'm getting two different results despite running the same version of the kernel as the source code. My distro may override the kernel's default setting, or maybe the value I found for `INOTIFY_WATCH_COST` is way off.

```console
$ cat /proc/sys/fs/inotify/max_user_watches
524288
$ ./a.out
watches_max: 1048576
```

In any event, I think setting this to `524288` is fine. I'm just curious to know how this value entered circulation. I'm also curious to see if we can prevent `tsc` from watching `node_modules` as I think we only care about changes to our application sources.

rbreslow reviewed (commented):
Everything looks great, Terrence. I'm sorry that I haven't submitted my feedback until now. I followed all your testing instructions, exercised the STRTA, and verified that `scripts/update` is _much faster_ on macOS. I left a few comments and have a few questions. I'm also looking forward to seeing how you update the README.

rajadain commented on Vagrantfile:
```diff
@@ -0,0 +1,48 @@
+  config.vm.synced_folder "~/.aws", "/home/vagrant/.aws"
+
+  config.vm.provider :virtualbox do |vb|
+    vb.memory = 6144
```
Based on https://github.com/PublicMapping/districtbuilder/pull/729#issuecomment-832394834 and https://azavea.slack.com/archives/C6UD5FG3T/p1619809216252100?thread_ts=1619804676.249600&cid=C6UD5FG3T

rajadain commented on Vagrantfile:
```diff
@@ -0,0 +1,48 @@
+
+  config.vm.provision "ansible_local" do |ansible|
+    ansible.compatibility_mode = "2.0"
+    ansible.extra_vars = { ansible_python_interpreter: "/usr/bin/python3" }
+    ansible.install_mode = "pip_args_only"
+    ansible.pip_install_cmd = "sudo apt-get install -y python3-distutils && curl https://bootstrap.pypa.io/pip/get-pip.py | sudo python3"
```
Oh I didn't know about this! That's great. Will add a commit changing to Python 3 as the default.

rajadain replied:
Wait, I didn't understand the original comment, sorry.

In this case we're installing `distutils` as part of `pip_install_cmd`, and in the linked commit we install it as part of the shell provisioner. Everything else is effectively the same.

Since the Python 3 `distutils` need to be in place _before_ the `get-pip.py` can be executed, I like the way this is done here better. Since the shell block is written later, I assumed it would be executed afterwards.

rbreslow replied:
> Wait, I didn't understand the original comment, sorry.

I had intended to surface the first instance of using Python 3 as the default interpreter since it's a relatively new convention. 

> Since the Python 3 `distutils` need to be in place _before_ the `get-pip.py` can be executed, I like the way this is done here better. Since the shell block is written later, I assumed it would be executed afterwards.

100% agree.

rbreslow commented:
I see that there's the checklist item for the README. Once that's in, I'll give a 👍

rajadain commented:
Rebased on latest `develop`, modified the `setup` script to support Vagrant, added instructions to the README, updated the CHANGELOG. Ran through the new instructions and they worked on my machine. Ready for a final look.

rajadain commented on docker-compose.yml:
```diff
@@ -33,6 +33,7 @@ services:
     volumes:
       - ./src/server:/home/node/app/server
       - ./src/shared:/home/node/app/shared
+      - /var/cache/district-builder-server-node-modules:/home/node/app/server/node_modules
```
This could potentially break the Docker environment. Would love to get someone running on Linux test this too.

rbreslow replied:
I was able to run `update` and `test` on my Linux workstation. This is the directory listing for `/var/cache`:

```console
[algiers districtbuilder] (research/add-vagrant)$ ls -l /var/cache/ | grep district-builder
drwxr-xr-x - root  4 Jun 16:10 district-builder-client-node-modules
drwxr-xr-x - root  4 Jun 16:14 district-builder-manage-node-modules
drwxr-xr-x - root  4 Jun 16:13 district-builder-server-node-modules
```

rbreslow commented on README.md:
```diff
@@ -28,22 +28,57 @@ DistrictBuilder is web-based, open source software for collaborative redistricti
 The server backend will need this in order to access S3 assets.
 
-Run `scripts/setup` to prepare the development environment:
+### Linux vs macOS
```
I would consider updating the table of contents at the top of the README to include these new sections.

rbreslow commented on README.md:
```diff
@@ -28,22 +28,57 @@ DistrictBuilder is web-based, open source software for collaborative redistricti
-Run `scripts/setup` to prepare the development environment:
+### Linux vs macOS
+
+The Docker containers used in development work very well on Linux, but require an additional layer of translation when running on non-Linux hosts. In particular, there are significant file-watching costs, which result in high CPU usage on macOS. On macOS, it is more efficient to run the containers within a Linux VM created with Vagrant.
```
👍

rbreslow commented on scripts/setup:
```diff
@@ -8,14 +8,20 @@ fi
     echo -n \
-        "Usage: $(basename "$0")
-Setup the project's development environment.
+        "Usage: $(basename "$0") [--vagrant]
+Setup the project's Docker-based development environment.
+This is recommended for Linux hosts.
+
+Optionally use --vagrant to setup a Vagrant-based environment.
+This is recommended for macOS hosts.
```
Nice.

rbreslow commented on README.md:
````diff
@@ -28,22 +28,57 @@ DistrictBuilder is web-based, open source software for collaborative redistricti
 
 ### Hot Reloading 🔥
 
+Next, run `scripts/server` to start the application:
+
+```bash
+ $ ./scripts/server
+```
+
````
This seems out of place in this section. **Hot Reloading** was intended to describe the ability of the development environment to pick up changes. I'd consider structuring things in such a way that there's the default path (e.g., Linux) that walks you through `update` and `server` and then an aside describing the differences with Vagrant.

However, I don't want this piece of feedback to hold up the PR from being merged. The more I look at the README, the more (preexisting) redundancy I see, and there's a lot of work we could do to simplify things.

rajadain replied:
Let's defer restructuring the README for when we can give it proper attention.

rbreslow reviewed (approved):
👍

maurizi reviewed (approved):
Doesn't seem to have caused any regressions for me 👍

rajadain commented:
Updated table of contents, renamed **Linux vs macOS** heading to **Host Environments**.

Thank you all for your reviews! Will merge once green. Hope that this will come in handy for all macOS developers of this project.
